src/tests/functional/visit_tests.js

```diff
@@ -220,6 +221,55 @@ test("test Visit with network error", async ({ page }) => {
   await nextEventNamed(page, "turbo:fetch-request-error")
 })
 
+test("test Visit direction data attribute when clicking a link", async ({ page }) => {
+  await Promise.all([
+    waitUntilSelector(page, "[data-turbo-visit-direction='forward']")
+      .then(() => waitUntilSelector(page, "html:not([data-turbo-visit-direction])")),
+    page.click("#same-origin-link")
+  ])
```

Comment:
I think we can write these tests without relying on `Promise.all`:

```js
  page.click("#same-origin-link")

  await waitUntilSelector(page, "[data-turbo-visit-direction='forward']")
  await waitUntilNoSelector(page, "[data-turbo-visit-direction]")
```

Reply:
I think I tried this and I don't this it worked  because the attribute is added and removed quicker than the assertions can track.

If I recall correctly, in the time it takes for `page.click` to resolve, the attribute it already added and removed, so `waitUntilSelector(page, "[data-turbo-visit-direction='forward']")` never passes

Reply:
also line 226-227 ensures the attribute is added and removed in the correct order

Reply:
@domchristie what do you think of a660c7bf644daf2f29abfdc146eec2367698cf10?

The trick is that you don't `await` for the `page.click`, so it runs in parallel with the first `waitUntilSelector`. Then we `await` for the two assertions to ensure they happen one after the other.

Reply:
Huh, that works locally but seems very flaky on CI.

Reply:
I think our best option is to use the mutation logs so we are not so dependent on the exact assertion timing. With the mutation log even if the attribute is no longer present in the page, we can still assert that it was present at some point.

See https://github.com/hotwired/turbo/compare/2bc7c13c66b2a31c4c46c63819d4f7f052538e6f..5d4758969663e4633cf5c61eb9b2bb23cf0a77a9

Reply:
> I think our best option is to use the mutation logs

This looks good to me 👍

Reply:
Cool, @domchristie if you can push those changes to your branch I'll merge.

Reply:
@afcapel I will do (I’ve just moved house, and will have a reliable internet connection next week)

Reply:
> if you can push those changes to your branch I'll merge

@afcapel Done